### PR TITLE
feat: materials

### PR DESCRIPTION
**Pull request by HuanCheng65**
## Summary by CodeRabbit

- **New Features**
    - Introduced functionality for handling material uploads, retrieval, and deletion with authentication.
    - Added custom error classes for specific material-related error cases.
    - Implemented a NestJS module for managing file uploads with Multer integration.
    - Added end-to-end tests for material uploading and retrieval.
- **Bug Fixes**
    - Removed unnecessary imports in comment and groups controllers.
- **Chores**
    - Configured binary attributes in `.gitattributes` for specific directories.
    - Updated `sample.env` with configuration for material file uploads.
    - Enhanced GitHub workflows with FFmpeg installation steps for testing.

### CODE REVIEW

**Comment by andylizf** [comment]
@coderabbitai review

File: src/materials/materials.controller.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+          meta,
+        },
+      },
+    };
+  }
+  @Delete(':/materialId')
+  async deleteMaterial() //@Param('materialId') id: number,
+  {
+    return {
+      code: 204,
```
Comment (HuanCheng65) [inline src/materials/materials.controller.ts:76]: Make it respond with code 204 directly

File: src/materials/materials.service.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    const newMaterial = await this.prismaService.material.create({
+      data: {
+        url: file.destination,
+        type: parseMaterial(type),
+        name: file.filename,
+        meta,
+      },
+    });
+    return newMaterial.id;
+  }
```
[the remark this answers is not in the document]
Comment (nameisyui) [reply src/materials/materials.service.ts:118]: yes,please give me some assistance

[the remark this answers is not in the document]
Comment (nameisyui) [reply src/materials/materials.service.ts:118]: fine

File: package.json [hunk 3]
```diff
@@ -39,28 +39,34 @@
     "@prisma/client": "5.9.1",
     "@types/bcrypt": "^5.0.2",
     "@types/bcryptjs": "^2.4.6",
-    "@types/express": "^4.17.21",
     "ajv": "^8.12.0",
     "bcryptjs": "^2.4.3",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.0",
     "express": "^4.18.2",
+    "fluent-ffmpeg": "^2.1.2",
```
Comment (andylizf) [inline package.json:47]: Are you sure you need ffmpeg? This package requires binary ffmpeg executable as prerequisite, which means you need some extra effort like configuring in dockerfile. See if you can generate thumbnail, etc. without ffmpeg binary executable.

**Review by andylizf — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (nameisyui) [reply package.json:47]: I cannot find another package that works better for getting image video audio data......If you know one please let me know.

File: src/materials/materials.module.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  MimeTypeNotMatchError,
+} from './materials.error';
+import { MaterialsService } from './materials.service';
+@Module({
+  imports: [
+    MulterModule.register({
+      storage: diskStorage({
+        destination: (req, file, callback) => {
+          let uploadPath: string;
+          switch (req.body.type) {
```
Comment (andylizf) [inline src/materials/materials.module.ts:21]: use a map instead

File: src/materials/materials.module.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            .join('');*/
+          const randomName = uuidv4();
+          return callback(null, `${randomName}${extname(file.originalname)}`);
+        },
+      }),
+      limits: {
+        fileSize: 1024 * 1024 * 1024, //1GB
+      },
+      fileFilter(req, file, callback) {
+        switch (req.body.type) {
```
Comment (andylizf) [inline src/materials/materials.module.ts:55]: use a map instead

File: src/materials/materials.service.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+import { PrismaService } from '../common/prisma/prisma.service';
+import { parseMaterial } from './materials.enum';
+import { MaterialNotFoundError, MetaDataParseError } from './materials.error';
+@Injectable()
+export class MaterialsService {
+  constructor(private readonly prismaService: PrismaService) {}
+  async getImageMetadata(
+    filePath: string,
+  ): Promise<{ width: number; height: number }> {
+    //ffmpeg.setFfmpegPath('E:/ffmpeg-master-latest-win64-gpl/bin/ffmpeg.exe');
```
Comment (andylizf) [inline src/materials/materials.service.ts:13]: delete comments

File: src/materials/materials.service.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+      };
+    } else {
+      meta = {
+        size: file.size,
+        name: file.filename,
+        mime: file.mimetype,
+        expires: 0, // todo?
+      };
+    }
+    console.log(meta);
```
Comment (andylizf) [inline src/materials/materials.service.ts:108]: remove console.log for submission

File: src/materials/materials.service.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    const newMaterial = await this.prismaService.material.create({
+      data: {
+        url: file.destination,
+        type: parseMaterial(type),
```
Comment (andylizf) [inline src/materials/materials.service.ts:112]: use EnumDto instead. See #112

File: src/materials/DTO/get-material.dto.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import { material } from '@prisma/client';
+import { BaseRespondDto } from '../../common/DTO/base-respond.dto';
+
+export class GetMaterialRespondDto extends BaseRespondDto {
+  data: {
+    material: material;
```
Comment (andylizf) [inline src/materials/DTO/get-material.dto.ts:6]: Do not returns a entity in database to the client. Use the entity to construct a DTO, with only useful fields according to API.yml.

File: src/materials/materials.controller.ts [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+export class MaterialsController {
+  constructor(
+    private readonly materialsService: MaterialsService,
+    private readonly authService: AuthService,
+  ) {}
+
+  @Post()
+  @UseInterceptors(FileInterceptor('file'))
+  async uploadMaterial(
+    @Body() req: UploadMaterialRequestDto,
```
Comment (andylizf) [inline src/materials/materials.controller.ts:36]: See #112 `{ type }: UploadMaterialRequestDto`

File: src/materials/materials.service.ts [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+@Injectable()
+export class MaterialsService {
+  constructor(private readonly prismaService: PrismaService) {}
+  async getImageMetadata(
+    filePath: string,
+  ): Promise<{ width: number; height: number }> {
+    //ffmpeg.setFfmpegPath('E:/ffmpeg-master-latest-win64-gpl/bin/ffmpeg.exe');
+    //ffmpeg.setFfprobePath('E:/ffmpeg-master-latest-win64-gpl/bin/ffprobe.exe');
+    return new Promise((resolve) => {
+      ffmpeg.ffprobe(filePath, (err, metadata) => {
```
Comment (andylizf) [inline src/materials/materials.service.ts:16]: wrap ffprobe in `util.promisify` to use await

Comment (andylizf) [reply src/materials/materials.service.ts:16]: like
```typescript
  try {
    const metadata = await ffprobeAsync(filePath);
    const duration = metadata.format.duration;
    if (duration === undefined) {
      throw new Error('Unable to get audio duration.');
    }
    return duration;
  } catch (error) {
    throw new Error('Failed to analyze audio file: ' + error.message);
  }
```

File: src/materials/materials.service.ts [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        }
+      });
+    });
+  }
+  async uploadMaterial(
+    type: string,
+    file: Express.Multer.File,
+  ): Promise<number> {
+    let meta;
+    if (type === 'image') {
```
Comment (andylizf) [inline src/materials/materials.service.ts:77]: so many switches. consider utilizing polymorphic

**Review by andylizf — CHANGES_REQUESTED** [review]

File: src/materials/materials.controller.ts [hunk 13]
```diff
@@ -55,10 +55,10 @@ export class MaterialsController {
   @Get('/:materialId')
   async getMaterialDetail(
     @Param('materialId', ParseIntPipe) id: number,
-    @Query('fields') fields: string,
+    @Query('fields') fields: string = 'url,meta',
   ): Promise<GetMaterialRespondDto> {
-    fields; // ?
-    const material = await this.materialsService.getMaterial(id);
+    const fieldList = fields.split(',');
```
Comment (andylizf) [inline src/materials/materials.controller.ts:60]: For this use case, you can utilize `ParseArrayPipe` in NestJs.
```typescript
    @Query('fields', new ParseArrayPipe({ separator: ',' })) fields: string[],
```

**Review by andylizf — CHANGES_REQUESTED** [review]

**Comment by andylizf** [comment]
@corporal9736 Looks like we need to add ffmpeg to the dependencies...

**Comment by corporal9736** [comment]
> @corporal9736 Looks like we need to add ffmpeg to the dependencies...

rebase the code to current dev and I can just add that. (the Dockerfile and actions are not sync with current dev)

**Review by andylizf — CHANGES_REQUESTED** [review]

File: src/materials/materials.service.ts [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    });
+  }
+  async getImageMetadata(
+    filePath: string,
+  ): Promise<{ width: number; height: number }> {
+    const metadata = await this.ffprobeAsync(filePath);
+    const width = metadata.streams[0].width;
+    const height = metadata.streams[0].height;
+    if (!width || !height) {
+      throw new MetaDataParseError('image');
```
Comment (andylizf) [inline src/materials/materials.service.ts:32]: Add comments to omit the warning from jest coverage reporter.

**Review by andylizf — CHANGES_REQUESTED** [review]

File: src/materials/materials.service.ts [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+      meta = {
+        size: file.size,
+        name: file.filename,
+        mime: file.mimetype,
+        expires: 0, // todo?
+      };
+    }
+    const newMaterial = await this.prismaService.material.create({
+      data: {
+        url: '/static/' + type + 's/' + file.filename, //todo:may need a better way
```
[the remark this answers is not in the document]
Comment (nameisyui) [reply src/materials/materials.service.ts:117]: Give me an example of baseStatic please.

[the remark this answers is not in the document]
Comment (nameisyui) [reply src/materials/materials.service.ts:117]: But we don't have a domain name yet.Because it's under development right now.

**Review by andylizf — APPROVED** [review]

**Pull request merged**